content/en/docs/ops/configuration/extensibility/wasm-module-distribution/index.md

```diff
@@ -0,0 +1,148 @@
+description: .
+weight: 10
+aliases:
+  - /help/ops/extensibility/distribute-remote-wasm-module
+  - /docs/ops/extensibility/distribute-remote-wasm-module
+keywords: [extensibility,Wasm,WebAssembly]
+owner: istio/wg-policies-and-telemetry-maintainers
+test: n/a
+---
+
+Istio provides the ability to [extend proxy functionality using WebAssembly (Wasm)](/blog/2020/wasm-announce/).
+One of the key advantages of proxy Wasm extensibility is that the extension can be loaded dynamically at runtime.
+Before loading, the Wasm extension needs to be distributed to the proxy.
+Istio provides a way to achieve this by downloading the Wasm module at Istio agent.
+
+## Configure a HTTP Filter with Remote Wasm Module
+
+Here we will walk through an example of adding a basic auth extension to our mesh. We will configure Istio to pull a [basic auth module](https://github.com/istio-ecosystem/wasm-extensions/tree/master/extensions/basic_auth) from a remote URI and load it with configuration to protect the `/productpage` URL.
+
+To configure a WebAssembly filter with a remote Wasm module, two `EnvoyFilter` resources will be installed: one injects the HTTP filter, and the other one provides configuration for the filter which uses the remote Wasm module.
+
+With the first `EnvoyFilter`, an HTTP filter will be injected into gateway proxies. It is configured to request the extension configuration named as `istio.basic_auth` from `ads` (i.e. Aggregated Discovery Service), which is the same configuration source that Istiod uses to provide all other configuration resources. Along with the configuration source, the initial fetch timeout is also set, in order to prevent filters with slow or failed Wasm module remote fetch becomes effective.
+
+{{< text yaml >}}
+apiVersion: networking.istio.io/v1alpha3
+kind: EnvoyFilter
+metadata:
+ name: basic-auth
+ namespace: istio-system
+spec:
+ configPatches:
+ - applyTo: HTTP_FILTER
+   match:
+     context: GATEWAY
+     listener:
+       filterChain:
+         filter:
+           name: envoy.http_connection_manager
+   patch:
+     operation: INSERT_BEFORE
+     value:
+       name: istio.basic_auth
+       config_discovery:
+         config_source:
+           ads: {}
+           initial_fetch_timeout: 0s # wait indefinitely to prevent bad Wasm fetch
+         type_urls: [ "type.googleapis.com/envoy.extensions.filters.http.wasm.v3.Wasm"]
+{{< /text >}}
+
+The second `EnvoyFilter` resource provides configuration for the filter, which is composed as an `EXTENSION_CONFIG` patch and will be distributed to the proxy as an Envoy [`Extension Configuration`](https://www.envoyproxy.io/docs/envoy/latest/configuration/overview/extension) (ECDS) resource.
+Once this update reaches Istio agent, it will download the Wasm module file and store it at the local file system.
+If the download fails, Istio-agent will reject the `Extension Configuration` update and prevent bad Wasm filter configuration from reaching Envoy.
```

Comment:
```suggestion
If the download fails, the agent will reject the `Extension Configuration` update to prevent invalid Wasm filter configuration from reaching the Envoy proxy.
```